lib/Auth/Process/RestoreAcrs.php

```diff
@@ -0,0 +1,79 @@
+ * IdP. The Perun Disco page, tries to add new ACRs to the request, if MFA has been requested, to prevent the IdP from
+ * failing. As a result, we then get modified requested ACRs, which should be restored to the previous (original) state
+ * using this authproc filter. It should be run on one of the first places of the IdP authproc chain.
+ */
+class RestoreAcrs extends ProcessingFilter
+{
+    public const CONFIG_FILE_NAME = 'module_perun.php';
+
+    private const DEBUG_PREFIX = 'perun:RestoreAcrs';
+
+    public function __construct($config, $reserved)
+    {
+        parent::__construct($config, $reserved);
+    }
+
+    public function process(&$request)
+    {
+        $this->restoreAcrs($request);
+    }
+
+    public static function storeAcrs(array &$state, array $acrsToAdd)
+    {
+        if (!empty($acrsToAdd)
+            && !empty($state[Disco::SAML_REQUESTED_AUTHN_CONTEXT][Disco::STATE_AUTHN_CONTEXT_CLASS_REF])
+            && sizeof($state[Disco::SAML_REQUESTED_AUTHN_CONTEXT][Disco::STATE_AUTHN_CONTEXT_CLASS_REF]) <= 1
+            && in_array(
+                Disco::MFA_PROFILE,
+                $state[Disco::SAML_REQUESTED_AUTHN_CONTEXT][Disco::STATE_AUTHN_CONTEXT_CLASS_REF],
+                true
+            )
+        ) {
+            Logger::debug(
+                self::DEBUG_PREFIX . ': Modifying ACRs list to pass regular authentication to the IdP to fallback to proxy MFA'
+            );
+            $original = [];
+            $new = [];
+            foreach ($state[Disco::SAML_REQUESTED_AUTHN_CONTEXT][Disco::STATE_AUTHN_CONTEXT_CLASS_REF] as $acr) {
+                $original[] = $acr;
+                $new[] = $acr;
+            }
+            foreach ($acrsToAdd as $acr) {
+                $new[] = $acr;
+            }
+            $new = array_unique($new);
+
+            unset($state[Disco::SAML_REQUESTED_AUTHN_CONTEXT][Disco::STATE_AUTHN_CONTEXT_CLASS_REF]);
+            if (isset($state[Disco::SAML_REQUESTED_AUTHN_CONTEXT_ORIGINAL])) {
+                unset($state[Disco::SAML_REQUESTED_AUTHN_CONTEXT_ORIGINAL]);
+            }
+            Logger::debug(self::DEBUG_PREFIX . ': original ACRs: ' . join(',', $original));
+            $state[Disco::SAML_REQUESTED_AUTHN_CONTEXT_ORIGINAL] = $original;
+            Logger::debug(self::DEBUG_PREFIX . ': ACRs after modification: ' . join(',', $new));
+            $state[Disco::SAML_REQUESTED_AUTHN_CONTEXT][Disco::STATE_AUTHN_CONTEXT_CLASS_REF] = $new;
+        }
+    }
+
+    private function restoreAcrs(&$request)
+    {
+        if (!empty($request[Disco::SAML_REQUESTED_AUTHN_CONTEXT_ORIGINAL])) {
+            unset($request[Disco::SAML_REQUESTED_AUTHN_CONTEXT][Disco::STATE_AUTHN_CONTEXT_CLASS_REF]);
+            $handle = &$request[Disco::SAML_REQUESTED_AUTHN_CONTEXT][Disco::STATE_AUTHN_CONTEXT_CLASS_REF];
+            $handle = $request[Disco::SAML_REQUESTED_AUTHN_CONTEXT_ORIGINAL];
```

Comment:
I don't think this is right. You override `$handle` here without using the first value. Also, the `&` in `&$request` should not be there, it is in the function signature.